Make Emacs still interactable while waiting for external quelpa processes

kiennq opened the pull request:
Another stab to try with running `quelpa-build--run-process` and still let the Emacs responsive.
@jcs090218, if you can try it out it would be great.

This is done based on using create a new edit loop to let the Emacs handle the keyboard while waiting for process output. Which is smaller than convert everything to async.

Fix #197

alphapapa commented:
I don't think this is a viable approach.  It's unclear what the implications are of using a recursive edit this way.  AFAIK it would be a misuse of it to do so.  I can imagine all sorts of weird behavior happening, which would hard for users to understand, as well as developers.  It would likely introduce many weird bugs.

A good question to ask when considering changes like this is, "What would emacs-devel say if this were proposed for installing into Emacs?"  I feel confident that this would be rejected on principle.

If a user needs Quelpa-like behavior with asynchronous implementation, it would probably be best to just use something designed with that in mind, like Elpaca.  We should make sure that Quelpa continues "just working" as it currently does, without risking the introduction of significant regressions.

As well, Emacs now has `package-vc`, which ideally should be able to work with async package installation (I don't know if it currently does).

steckerhalter commented:
@kiennq does that mean when quelpa runs it also handles keyboard input so it kinda takes over the input handling from Emacs?

kiennq commented:
> @kiennq does that mean when quelpa runs it also handles keyboard input so it kinda takes over the input handling from Emacs?

Yes, it will still handle keyboard so that the user can still insert/modify buffer. The only time when it stops response to input is after the package has been fetch and starts installing.

I have another approach of using emacs thread which should be more reliable than this, but that requires emacs 26. Using `recursive-edit` is fragile, but it __kinda__ works under moderate workload as I have tried so far. It would be great if someone can try this and give some feed backs.
I can push the thread approach here as well if you want to try.

steckerhalter commented:
I think the keyboard thing could cause unforseen issues. not sure about the thread approach. how about making it truly async? would that be possible?

alphapapa commented:
> Using `recursive-edit` is fragile, but it **kinda** works under moderate workload as I have tried so far. 

"Kinda works" would not be acceptable.

Emacs threads are also unsuitable for this task (they are generally not suitable for much of anything, sadly, due to their implementation).

If you want to run external processes asynchronously, you should look at the `async` library and how it implements asynchronous package installation and building.

The ideas in this PR are simply no-gos.  Quelpa is not a playground for experimental code and wacky hacks.  It is a tool relied upon by users to get work done.  

I'm afraid I have to say that this PR should not even have been suggested.  @steckerhalter The fact that it was calls into question again the submitter's technical judgment.  I'm not sure that someone who thinks this would be an appropriate change to make should have commit access to part of a toolchain used to build and install software for users around the world.

steckerhalter commented:
@alphapapa you could try to be a bit nicer. if @kiennq likes to try to make quelpa async that is a good intention. 

I agree though that the implementation should be using the async framework from Emacs. And if that is too difficult but @kiennq and anyone else likes to use his custom async approach, it can be made available optionally so it has to be enabled by the user. This way it doesn't disrupt anyone using quelpa.

So this would be my votum. If you can make quelpa async with the standard async framework and it is found to be stable and reliable a switch to this code could be made. Any other approach outside of the async framework could be added but not enabled by default, so the user has to enable the code and is aware of possible issues... what do you think?

kiennq commented:
> Emacs threads are also unsuitable for this task (they are generally not suitable for much of anything, sadly, due to their implementation).

You should suggest that to emacs-devel, I'm sure they will have different opinions on that. The thread using on this approach yield back to main thread when it's waiting for external process. That allows event in the main thread still be able to be processed while keeps the current code structure. Moving the whole thing to callbacks / async (not that async library which run thing in other process, Emacs has generator which supports coroutine async) is a lot more work (@steckerhalter ).

> If you want to run external processes asynchronously, you should look at the async library and how it implements asynchronous package installation and building.

I had an implementation at #227, that uses `async` before, upgrading a package in another process and callback into the current one is not working out well, mainly because the internal state for package from other process is not replicated nicely in the current one, you will likely end up with broken package somewhere down the line eventually.

> The ideas in this PR are simply no-gos. Quelpa is not a playground for experimental code and wacky hacks. It is a tool relied upon by users to get work done.

Experimental code can be hidden behind a flag, that's how software development has been done for decades. I'm looking for feedbacks with the initial implementation so I put that without any and anyone for like to participate in improving `Quelpa` can try.

> I'm afraid I have to say that this PR should not even have been suggested. @steckerhalter The fact that it was calls into question again the submitter's technical judgment. I'm not sure that someone who thinks this would be an appropriate change to make should have commit access to part of a toolchain used to build and install software for users around the world.

Do you have any grudge against me or something? It feels like when I'm trying to improve the current behavior with my limited spare time, you just try to dismiss it with your unconstructive comments. Have you tried the `quelpa-upgrade-all` command that locks up Emacs for a long time while waiting for package downloading? That part can be improved even with a single thread on a single CPU nature of Emacs.

> So this would be my votum. If you can make quelpa async with the standard async framework and it is found to be stable and reliable a switch to this code could be made. Any other approach outside of the async framework could be added but not enabled by default, so the user has to enable the code and is aware of possible issues... what do you think?

With the latest change I pushed, the new functionality needs to be invoked via `quelpa-xxx-async` command already.

steckerhalter commented:
@kiennq sounds good. do you think it would make sense to add a flag which allows one to switch to async? something like `quelpa-async-p` which defaults to `nil` and when you set it `t` all commands would run async? I think this would be quite nice and to have a build with the flag enabled would give you some indication if it is working well or not.
I rarely even run upgrades and so on but I guess could try it out like this as well and give some feedback

@alphapapa I'd appreciate if you could make constructive comments instead of discouraging development altogether. we can find a way to add features without having to risk breaking quelpa and if the async code proves to be reliable we can think about enabling it generally at some later point.

alphapapa commented:
> @alphapapa you could try to be a bit nicer. if @kiennq likes to try to make quelpa async that is a good intention.

I drafted it several times before submitting.  I'm afraid that's as nicely as I could find a way to put it.

We don't take supply-chain issues seriously enough in the Emacs community.  This is a good example of it.  Those who are entrusted with the power to push commits that will end up on users' machines, and especially commits that end up installing other software, should be well-known, with a strong reputation for technical excellence and cooperation.  I'm afraid that this contributor does not meet that bar, and this PR is further evidence of it.

> 
> I agree though that the implementation should be using the async framework from Emacs. And if that is too difficult but @kiennq and anyone else likes to use his custom async approach, it can be made available optionally so it has to be enabled by the user. This way it doesn't disrupt anyone using quelpa.

I strongly disagree.  This whole approach is nonsensical.  It exhibits a lack of basic understanding of Emacs's paradigms, as well as the habits of users.  It should never be a part of Quelpa.  If @kiennq wants to use it, it should reside in a fork.

alphapapa commented:
> > "Kinda works" would not be acceptable.
> 
> The test passed for the approach that uses `recursive-edit`, so the users can always just not touch Emacs (like how they have been forced to do before) when install package via Quelpa and nothing should be broken. 

Using recursive edits like this is a recipe for user confusion, frustration, and potentially disaster.  "Just don't touch Emacs" is not an acceptable user interaction paradigm.


> > Emacs threads are also unsuitable for this task (they are generally not suitable for much of anything, sadly, due to their implementation).
> 
> You should suggest that to emacs-devel, I'm sure they will have different opinions on that. 

I am speaking from experience, having worked with Emacs threads myself, and having participated in threads-related discussions on emacs-devel for years.  You don't seem to even be aware of the serious drawbacks involved, which is why they have yet to be used for any serious application.  Quelpa should not be the first.

> The thread using on this approach yield back to main thread when it's waiting for external process. That allows event in the main thread still be able to be processed while keeps the current code structure. Moving the whole thing to callbacks / async (not that async library which run thing in other process, Emacs has generator which supports coroutine async) is a lot more work (@steckerhalter ).

I am aware of how `generator.el` works.  It is very cool.  It is also complicated.  It makes debugging much more difficult.  In the same way as with threads, there are few-to-no serious applications of it, because of the inherent drawbacks.  Quelpa should not be the first.

> > If you want to run external processes asynchronously, you should look at the async library and how it implements asynchronous package installation and building.
> 
> I had an implementation at #227, that uses `async` before, upgrading a package in another process and callback into the current one is not working out well, mainly because the internal state for package from other process is not replicated nicely in the current one, you will likely end up with broken package somewhere down the line eventually.

The reverse is true: compiling packages in external processes guarantees a clean environment, not polluted by the user's, which can cause and obscure bugs.  Compiling packages outside of the main Emacs process is a best-practice.  This is why it's done in other tools like Elpaca, and my own https://github.com/alphapapa/makem.sh

> 
> > The ideas in this PR are simply no-gos. Quelpa is not a playground for experimental code and wacky hacks. It is a tool relied upon by users to get work done.
> 
> Experimental code can be hidden behind a flag, that's how software development has been done for decades. I'm looking for feedbacks with the initial implementation so I put that without any and anyone for like to participate in improving `Quelpa` can try.

This is a bad idea.  If you want to work on it, it's your time, but it should be in a fork.

> 
> > I'm afraid I have to say that this PR should not even have been suggested. @steckerhalter The fact that it was calls into question again the submitter's technical judgment. I'm not sure that someone who thinks this would be an appropriate change to make should have commit access to part of a toolchain used to build and install software for users around the world.
> 
> Do you have any grudge against me or something? It feels like when I'm trying to improve the current behavior with my limited spare time, you just try to dismiss it with your unconstructive comments. Have you tried the `quelpa-upgrade-all` command that locks up Emacs for a long time while waiting for package downloading? That part can be improved even with a single thread on a single CPU nature of Emacs.

Every argument I've made has been based on sound technical reasoning.  You have not answered any of those arguments with sound technical rebuttals.  Projects like this should pursue technical excellence and put egos aside.

steckerhalter commented:
@alphapapa you continue with this mean-spirited attitude. this is not a professional project and it suffices that we have to deal with this kind of mindset at workplace sometimes. if you have a better implementation then you are free to contribute but stop putting others down for trying to improve quelpa. I will not tolerate this any longer. and don't just close issues as you were in charge. I prefer people like @kiennq any day who actually do something. all you can do is criticize but you don't write a line of code.

kiennq commented:
> @kiennq sounds good. do you think it would make sense to add a flag which allows one to switch to async? something like `quelpa-async-p` which defaults to `nil` and when you set it `t` all commands would run async? I think this would be quite nice and to have a build with the flag enabled would give you some indication if it is working well or not.
> I rarely even run upgrades and so on but I guess could try it out like this as well and give some feedback

Done, I've added `quelpa-async-p`.
Also, I reverted back to use `recursive-edit` instead of making a new thread. The reason is because making a new thread will just return immediately, so if we run it in a batch environment where we want to wait for the previous packages to be installed before the next package can be proceeded, we will need to block the thread until it finishes. That will in turn block the input.
Another reason is the `recursive-edit` approach is much cleaner and I have been using it for sometimes, even when bootstrap a new Emacs environment with packages downloading via both Quelpa and `package.el` so I'm more confident of its usability.
I guess if I try messing around more when the Emacs is in recursive edit state, something can break. But it would be easier recovered via `top-level` to exit all recursive edits and rerun the quelpa command again.
@steckerhalter please give it a try if you have time.

steckerhalter commented:
@kiennq thanks. how about the failing builds? can you either remove or fix them?

jcs090218 commented:
I'm sorry for not getting back to you sooner. This somehow goes under my radar. 🤔 

I've tested it, and it is working correctly! 🎉   Thank you for working on this! 🚀

alphapapa commented:
> @alphapapa you continue with this mean-spirited attitude.

No, I'm about as far from that as you can get.  What I am doing, though, is trying to put the interests of users first, even if that means having to criticize someone's code.  What is the point of a project like this, if not to be a useful, reliable tool for other people?

>  stop putting others down for trying to improve quelpa.   I will not tolerate this any longer. 

It's not my goal to put anyone down.  Sometimes, though, a firm stance is needed when unwise changes are proposed, especially if the one proposing them is unreceptive to feedback from those who have some experience in the matter.

>and don't just close issues as you were in charge. 

I have not closed this issue.  

I did recently close https://github.com/quelpa/quelpa/issues/255#issuecomment-2395819007, because [the backtrace](https://github.com/quelpa/quelpa/issues/255#issuecomment-2395508934) shows that the error is not coming from Quelpa, so it appears to not be a bug in Quelpa.

As mentioned on https://github.com/quelpa/quelpa/issues/252, four years ago you asked for someone to take over maintenance of Quelpa, and I volunteered.  You then accepted my offer, as well as bringing kiennq and jcs on board.  You then decided to not step down yourself, but to maintain it as a team.

Now you suddenly say that I'm not allowed to close bug reports that appear to be invalid.  I don't know what to say to that.

> I prefer people like @kiennq any day who actually do something. all you can do is criticize but you don't write a line of code.

I don't think that's a fair assessment.

Regardless, it's clear that our ideas of this project's purpose and development methodology are incompatible.  So I will cheerfully disassociate myself from it now.  I wish you all continued success in Quelpa's maintenance and development.

steckerhalter commented:
> What I am doing, though, is trying to put the interests of users first, even if that means having to criticize someone's code. What is the point of a project like this, if not to be a useful, reliable tool for other people?

async is very useful. with your behavior you discourage the implementation. your arguments are meaningless because it doesn't change anything regarding the realiability since the default is not changing. and you don't even know if ithe async implementation is not reliable since you have never tested it.

now stop distracting the people who actually make a difference.

jcs090218 commented:
> @kiennq thanks. how about the failing builds? can you either remove or fix them?

The CI is broken because the `run-stuff` package no longer supports 27.x and 28.x. 🤔  See https://codeberg.org/ideasman42/emacs-run-stuff/src/branch/main/run-stuff.el#L11.

We can simply remove the test under 27 and 28 or test with another package in Codeberg. :)

kiennq commented:
> > @kiennq thanks. how about the failing builds? can you either remove or fix them?
> 
> The CI is broken because the `run-stuff` package no longer supports 27.x and 28.x. 🤔 See https://codeberg.org/ideasman42/emacs-run-stuff/src/branch/main/run-stuff.el#L11.
> 
> We can simply remove the test under 27 and 28 or test with another package in Codeberg. :)

Thanks, should be fixed now.

steckerhalter commented:
I haven't tested it yet. Are you confident that it works well in the current state?

kiennq commented:
> I haven't tested it yet. Are you confident that it works well in the current state?

It should be.

jcs090218 commented:
Might it be a good idea to add async tests as well? 🤔

steckerhalter commented:
@jcs090218 yes that would help

kiennq commented:
Added async test with `quelpa-async-p` set to `t`.
Right now source hut test is failing due to server being down or something, all other tests are passing.

kiennq commented:
@jcs090218 @steckerhalter is it okay for me to merge this now?

jcs090218 commented:
LGTM! I am not sure why the CI is failing, though. 🤔

steckerhalter commented:
@kiennq it's ok to merge for me. would be good if you could make the CI work before merging. cheers

kiennq commented:
There's one failure in connecting to emacs-wiki now but it passed in other versions so a transient issue.

fnussbaum commented:
> you are the one who wanted to move the whole project from the quelpa org to your space for no good reason. this would have broken quelpa for EVERY USER.

In case you would like to reconsider at some point in the future: repository transfers can be quite seamless as far as I know. I don't think this would necessarily have broken anything.

steckerhalter commented:
@fnussbaum I don't think it would habe been seamless. every quelpa user has the github url coded into their .emacs
I think it would have broken quelpa for every user and every one of them would have had to adjust their .emacs to fix it.
I'm not talking about the transfer on github. also why would you want to abandon the quelpa org and move it to a personal namespace? this doesn't make sense to me

fnussbaum commented:
@steckerhalter I recently looked into this because we had a similar request to move Spacemacs to an organization. And it seems as if Github would automatically install all kinds of redirects, such that one could even still `git pull` from the old URL, for example. To be fair, I have not actually tried it though.

I guess it does not matter now for quelpa, as it is already hosted by an organization.

steckerhalter commented:
@fnussbaum ok hmm, I didn't know this is possible. I wouldn't have a good feeling about it though :smirk: and since it's already an org there is no good reason to do this IMO
